nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java

```diff
@@ -171,6 +204,516 @@ public Response getLoginConfig(@Context HttpServletRequest httpServletRequest) {
+            return;
+        }
+
+        // ensure a RelayState value was sent back
+        final String requestState = parameters.get("RelayState");
+        if (requestState == null) {
+            removeSamlRequestCookie(httpServletResponse);
+            forwardToLoginMessagePage(httpServletRequest, httpServletResponse, "The RelayState parameter was not found in the request. Unable to continue.");
+            return;
+        }
+
+        // ensure the RelayState value in the request matches the store state
+        if (!samlStateManager.isStateValid(samlRequestIdentifier, requestState)) {
+            logger.error("The RelayState value returned by the SAML IDP does not match the stored state. Unable to continue login process.");
+            removeSamlRequestCookie(httpServletResponse);
+            forwardToLoginMessagePage(httpServletRequest, httpServletResponse, "Purposed RelayState does not match the stored state. Unable to continue login process.");
+            return;
+        }
+
+        // process the SAML response
+        final SAMLCredential samlCredential;
+        try {
+            samlCredential = samlService.processLogin(httpServletRequest, httpServletResponse, parameters);
+        } catch (Exception e) {
+            removeSamlRequestCookie(httpServletResponse);
+            forwardToLoginMessagePage(httpServletRequest, httpServletResponse, e.getMessage());
+            return;
+        }
+
+        // create the login token
+        final String rawIdentity = samlCredential.getNameID().getValue();
+        final String mappedIdentity = IdentityMappingUtil.mapIdentity(rawIdentity, IdentityMappingUtil.getIdentityMappings(properties));
+        final long expiration = validateTokenExpiration(samlService.getAuthExpiration(), mappedIdentity);
+        final String issuer = samlCredential.getRemoteEntityID();
+
+        final LoginAuthenticationToken loginToken = new LoginAuthenticationToken(mappedIdentity, mappedIdentity, expiration, issuer);
+
+        // create and cache a NiFi JWT that can be retrieved later from the exchange end-point
+        samlStateManager.createJwt(samlRequestIdentifier, loginToken);
+
+        // store the SAMLCredential for retrieval during logout
+        // issue a delete first in case the user already had a stored credential that somehow wasn't properly cleaned up on logout
+        samlCredentialStore.delete(mappedIdentity);
+        samlCredentialStore.save(mappedIdentity, samlCredential);
+
+        // get the user's groups from the assertions if the exist and store them for later retrieval
+        final Set<String> userGroups = samlService.getUserGroups(samlCredential);
+        storeIdpGroups(mappedIdentity, IdpType.SAML, userGroups);
+
+        // redirect to the name page
+        httpServletResponse.sendRedirect(getNiFiUri());
+    }
+
+    private void storeIdpGroups(final String userIdentity, final IdpType idpType, final Set<String> groupNames) {
```

Comment:
nit-pick... Should the logic in `samlLoginConsumer`, `storeIdpGroups`, and `samlSingleLogoutConsumer` be elsewhere? Seems like a lot to have in the `AccessResource`.

Reply:
I'll see if it makes sense to push some of this logic down into the service.